File: javascript/src/Creational/Abstract_Factory/AbstractSoupFactory.js
(function() {
  define([], function() {
    var AbstractSoupFactory;
    AbstractSoupFactory = (function() {
      function AbstractSoupFactory(factoryLocation) {
        this.factoryLocation = factoryLocation;
        throw "Can't instantiate an abstract class, AbstractSoupFactory";
      }

      AbstractSoupFactory.prototype.getFactoryLocation = function() {
        return this.factoryLocation;
      };

      AbstractSoupFactory.prototype.makeChickenSoup = function() {
        return new ChickenSoup();
      };

      AbstractSoupFactory.prototype.makeClamChowder = function() {
        return new ClamChowder();
      };

      AbstractSoupFactory.prototype.makeFishChowder = function() {
        return new FishChowder();
      };

      AbstractSoupFactory.prototype.makeMinnestrone = function() {
        return new Minnestrone();
      };

      AbstractSoupFactory.prototype.makePastaFazul = function() {
        return new PastaFazul();
      };

      AbstractSoupFactory.prototype.makeTofuSoup = function() {
        return new TofuSoup();
      };

      AbstractSoupFactory.prototype.makeVegetableSoup = function() {
        return new VegetableSoup();
      };

      return AbstractSoupFactory;

    })();
    return AbstractSoupFactory;
  });

}).call(this);

File: csharp/src/Creational/Abstract_Factory/AbstractSoupFactory.cs

//// Original copy of this content taken from http://www.fluffycat.com/Java-Design-Patterns/ in 2010
//// Original Author: Larry Truett
//// Privacy Policy at http://www.fluffycat.com/Privacy-Policy/
//An abstract factory has sets of methods to make families of various objects.

//In this example the AbstractSoupFactory defines the method names and return types to make various kinds of soup.

//The BostonConcreteSoupFactory and the HonoluluConcreteSoupFactory both extend the AbstractSoupFactory.

//An object can be defined as an AbstractSoupFactory, and instantiated as either a BostonConcreteSoupFactory (BCSF) or a HonoluluConcreteSoupFactory (HCSF). Both BCSF or HCSF have the makeFishChowder method, and both return a FishChowder type class. However, the BCSF returns a FishChowder subclass of BostonFishChowder, while the HCSF returns a FishChowder subclass of HonoluluFishChowder.
//Still reading? Save your time, watch the video lessons!
//Video tutorial on design patterns
//AbstractSoupFactory.java - An Abstract Factory


namespace Creational.Abstract_Factory {

using common;

public abstract class AbstractSoupFactory {

    protected string factoryLocation;

    public string getFactoryLocation() {
        return factoryLocation;
    }
    
    public ChickenSoup makeChickenSoup() {
        return new ChickenSoup();
    }
    public ClamChowder makeClamChowder() {
        return new ClamChowder();
    }
    public FishChowder makeFishChowder() {
        return new FishChowder();
    }     
    public Minnestrone makeMinnestrone() {
        return new Minnestrone();
    }
    public PastaFazul makePastaFazul() {
        return new PastaFazul();
    }
    public TofuSoup makeTofuSoup() {
        return new TofuSoup();
    }
    public VegetableSoup makeVegetableSoup() {
        return new VegetableSoup();
    }
}

}
